### PR TITLE
feat: add Debridge protocol implementation with transaction parsing

### LINKED ISSUE #27: Integrate Debridge
https://docs.debridge.finance/

### PR DESCRIPTION
**Pull request by Subham-Satapathy**
## Description

Fixes #27 

## Type of change

- [ ] Updated dependencies.
- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Is this a Breaking change? (fix or feature that would cause existing functionality to not work as expected)
- [ ] Did this get a semantic version bump?
- [x] Does this PR include any Tests for the changes made or already existing codebase?
- [ ] Documentation updates.
- [x] Configuration changes.

## Checklist

- [x] Self-review completed
- [x] Tests passing
- [ ] Branch synced with base branch
- [x] Code follows project style guidelines
- [x] Performance implications considered
- [ ] Security implications considered
- [x] No sensitive information included
- [x] Version bumped appropriately

## Documentation

- [ ] README updated if needed.
- [ ] API documentation updated.
- [ ] Comments added to complex code.
- [ ] Change log updated.

### CODE REVIEW

File: src/lib/Debridge/contracts.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  // Event for when a order is placed
+  ORDER_PLACED = "0xfc8703fd57380f9dd234a89dce51333782d49c5902f307b02f03e014d18fe471",
+  // Event for when a order is fulfilled
+  ORDER_FULFILLED = "0xd281ee92bab1446041582480d2c0a9dc91f855386bb27ea295faac1e992f7fe4"
+}
+
+export const contracts: IProtocolContractDefinitions = {
+  [CONTRACT_ENUM.DLN_SOURCE]: {
+    interface: new ethers.Interface(DlnSourceAbi),
+    deployments: {
```
Comment (divyesh87) [inline src/lib/Debridge/contracts.ts:23]: Please add all the deployment addresses on all the chains. You can use an explorer like routescan that can give you all the transactions for all the chains

https://routescan.io/address/0xeF4fB24aD0916217251F553c0596F8Edc630EB66
Screenshot 2024-12-25 at 05 38 05

File: src/lib/Debridge/contracts.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    events: {
+      [EVENT_ENUM.ORDER_PLACED]: {
+        abi: new ethers.Interface(["event CreatedOrder(tuple(uint64 makerOrderNonce, bytes makerSrc, uint256 giveChainId, bytes giveTokenAddress, uint256 giveAmount, uint256 takeChainId, bytes takeTokenAddress, uint256 takeAmount, bytes receiverDst, bytes givePatchAuthoritySrc, bytes orderAuthorityAddressDst, bytes allowedTakerDst, bytes allowedCancelBeneficiarySrc, bytes externalCall) order, bytes32 orderId, bytes affiliateFee, uint256 nativeFixFee, uint256 percentFee, uint32 referralCode, bytes metadata)"])
+      }
+    }
+
+  },
+  [CONTRACT_ENUM.DLN_DESTINATION]: {
+    interface: new ethers.Interface(DlnDestinationAbi),
+    deployments: {
```
Comment (divyesh87) [inline src/lib/Debridge/contracts.ts:54]: Same for this, make sure to cover every chain

File: src/lib/Debridge/index.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import { ProtocolHelper } from '../../helpers';
+import { IProtocolContractDefinitions, IProtocolParserExport, ITransaction, ITransactionAction } from '../../types';
+import { CONTRACT_ENUM, contracts } from './contracts';
+import { DlnSourceContractParseTransaction, DlnDestinationContractParseTransaction } from './parser';
+
+export default class Debridge implements IProtocolParserExport {
+  public readonly protocolIdentifier: string;
+
+  constructor() {
+    this.protocolIdentifier = protocols.debridge.identifier;
```
Comment (divyesh87) [inline src/lib/Debridge/index.ts:11]: Please define debridge under protocols and add the necessary info

File: src/lib/Debridge/parser.ts [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        toChain: order.takeChainId.toString(),
+        fromToken: order.giveTokenAddress,
+        toToken: order.takeTokenAddress,
+        fromAmount: (BigInt(order.giveAmount) + BigInt(structuredEvent.percentFee)).toString(),
+        toAmount: order.takeAmount.toString(),
+        sender: order.makerSrc,
+        recipient: order.receiverDst
+      };
+    } catch (error) {
+      console.error('Error parsing order placed transaction:', error);
```
Comment (divyesh87) [inline src/lib/Debridge/parser.ts:52]: Lets not have any console logs in the parser. You can just throw with an appropriate message

File: tests/lib/Debridge/parsedLog.json [hunk 5]
```diff
@@ -0,0 +1,226 @@
+{
```
Comment (divyesh87) [inline tests/lib/Debridge/parsedLog.json:1]: Whats the need for this file?

File: src/lib/Debridge/contracts.ts (hunk 1 above)
Comment (Subham-Satapathy) [reply src/lib/Debridge/contracts.ts:23]: Added

File: src/lib/Debridge/contracts.ts (hunk 2 above)
Comment (Subham-Satapathy) [reply src/lib/Debridge/contracts.ts:54]: Done

File: src/lib/Debridge/parser.ts (hunk 4 above)
Comment (Subham-Satapathy) [reply src/lib/Debridge/parser.ts:52]: Done

Comment (Subham-Satapathy) [reply src/lib/Debridge/parser.ts:52]: Let's discuss this

File: tests/lib/Debridge/parsedLog.json (hunk 5 above)
Comment (Subham-Satapathy) [reply tests/lib/Debridge/parsedLog.json:1]: Not needed. It's a test file

File: src/lib/Debridge/index.ts (hunk 3 above)
Comment (Subham-Satapathy) [reply src/lib/Debridge/index.ts:19]: Done

File: src/lib/Debridge/parser.ts (hunk 4 above)
Comment (Subham-Satapathy) [reply src/lib/Debridge/parser.ts]: Done

File: src/lib/Debridge/parser.ts [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    );
+    if (matchedPlacedOrderLog) {
+      actions.push(this.parsePlacedOrder(matchedPlacedOrderLog));
+    }
+    return actions;
+  }
+
+  private static parsePlacedOrder(
+    placeOrderLog: ITransactionLog
+  ): IBridgeInAction {
```
Comment (divyesh87) [inline src/lib/Debridge/parser.ts:37]: The dln source contract parser is for bridge out actions no?

File: src/lib/Debridge/parser.ts [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    const parsedLog = ProtocolHelper.parseLog(
+      placeOrderLog,
+      this.contractDefiniton.events[EVENT_ENUM.ORDER_PLACED]
+    );
+
+    const order = parsedLog.args.order;
+
+    return {
+      type: ACTION_ENUM.BRIDGE_IN,
```
Comment (divyesh87) [inline src/lib/Debridge/parser.ts:46]: For source contract this should be a bridge out action

File: src/lib/Debridge/parser.ts [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+      actions.push(this.parseClaimTransaction(transaction, parsedTxn));
+    }
+
+    return actions;
+  }
+
+  private static parseSendTransaction(
+    transaction: ITransaction,
+    parsedTxn: ethers.TransactionDescription
+  ): IBridgeInAction {
```
Comment (divyesh87) [inline src/lib/Debridge/parser.ts:126]: send should be bridge out

File: tests/lib/Debridge/data.ts [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+          toAmount: "2994184",
+          sender: "0x555ce236c0220695b68341bc48c68d52210cc35b",
+          recipient: "0x807f4f82f02f7c986a51887ee5536aed767977c8",
+        },
+      ],
+    },
+  ],
+};
+
+export const debridgeCrossChainOrderData: IProtocolTestingData = {
```
Comment (divyesh87) [inline tests/lib/Debridge/data.ts:249]: Lets try to add atleast 5 transactions for each contract!

File: src/lib/Debridge/contracts.ts [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+      },
+      [CHAIN_ID.SONIC]: {
+        address: "0xeF4fB24aD0916217251F553c0596F8Edc630EB66",
+        listenForTransactions: [LISTEN_FOR_TRANSACTIONS.INCOMING],
+      },
+      [CHAIN_ID.ZORA]: {
+        address: "0xeF4fB24aD0916217251F553c0596F8Edc630EB66",
+        listenForTransactions: [LISTEN_FOR_TRANSACTIONS.INCOMING],
+      },
+      [CHAIN_ID.METIS]: {
```
[the remark this answers is not in the document]
Comment (Subham-Satapathy) [reply src/lib/Debridge/contracts.ts:64]: Added chains

File: src/lib/Debridge/parser.ts [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+    const order = parsedLog.args.order;
+
+    return {
+      type: ACTION_ENUM.BRIDGE_IN,
+      fromChain: Number(order.giveChainId),
+      toChain: order.takeChainId.toString(),
+      fromToken: order.giveTokenAddress,
+      toToken: order.takeTokenAddress,
+      fromAmount: (
+        BigInt(order.giveAmount) + BigInt(parsedLog.args.percentFee)
```
[the remark this answers is not in the document]
Comment (Subham-Satapathy) [reply src/lib/Debridge/parser.ts:52]: Done

File: src/lib/Debridge/parser.ts [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+        toChain: transaction.chainId.toString(),
+        fromToken: contractAddress.toString(),
+        toToken: null,
+        fromAmount: value.toString(),
+        toAmount: null,
+        sender: transaction.from.toString(),
+        recipient: parsedTxn.args._receiver.toString(),
+      };
+    } else {
+      console.log("No ERC20 transfer logs found");
```
[the remark this answers is not in the document]
Comment (Subham-Satapathy) [reply src/lib/Debridge/parser.ts:161]: Done

File: src/lib/Debridge/parser.ts [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+  private static parseClaimTransaction(
+    transaction: ITransaction,
+    parsedTxn: ethers.TransactionDescription
+  ): IBridgeInAction {
+    const log = transaction.logs;
+
+    const parsedLogs = ProtocolHelper.parseERC20TransferLogs(log);
+    if (parsedLogs.length > 0) {
+      const { value, contractAddress } = parsedLogs[0];
```
[the remark this answers is not in the document]
Comment (Subham-Satapathy) [reply src/lib/Debridge/parser.ts:148]: Done

File: src/lib/Debridge/parser.ts (hunk 8 above)
Comment (Subham-Satapathy) [reply src/lib/Debridge/parser.ts:126]: Done

File: tests/lib/Debridge/data.ts (hunk 9 above)
Comment (Subham-Satapathy) [reply tests/lib/Debridge/data.ts:311]: Done

**Pull request merged**